core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -411,10 +411,9 @@ class KafkaApis(val requestChannel: RequestChannel,
         }
 
         val topicPartition = new TopicPartition(topicName, partition.index())
-        if (topicName.isEmpty)
+        // To compatible with the old version, only return UNKNOWN_TOPIC_ID if topicId is not default value
+        if (topicName.isEmpty && !topic.topicId().equals(Uuid.ZERO_UUID))
```

Comment:
If the request version is 13 and the topicId is set to 0, we should return UNKNOWN_TOPIC_ID to match what's in the fetch request. Only if the request version is below 13 and topicName is empty, should we return UNKNOWN_TOPIC_OR_PARTITION. So, it seems that we should do `fetchRequest.version() >= 13` instead of `!topic.topicId().equals(Uuid.ZERO_UUID)`.

Reply:
That makes sense. I left the similar comment before - https://github.com/apache/kafka/pull/19635#discussion_r2073937574